Repo URL: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted> \
Commit ID: ac5dab4 \
NOTE: This is technically the PREVIOUS commit id, as adding the Commit ID will cause a new commit, which would have a different ID, which I would then need to update, and so on. No code has been changed since the previous commit.

In [1]:
# Imports section
import pandas as pd
from sklearn import *
# just using for mean function
import statistics as s

## Part 1. Loading the dataset

In [53]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

sci_data = pd.read_csv("science_data_large.csv")
print(sci_data.head(15))
sci_data.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [54]:
# Take the pandas dataset and split it into our features (X) and label (y)



X = sci_data.iloc[:,0:2]
Y = sci_data.iloc[:,2]


# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, Y_train, Y_test = sklearn.model_selection.train_test_split(X, Y, test_size=0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [55]:
# Use sklearn to train a model on the training set
model = sklearn.linear_model.LinearRegression().fit(X_train, Y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_X = X_test.iloc[1:3,:]

#print(sample_X)
sample_prediction = model.predict(sample_X)
sample_score = model.score(sample_X, sample_prediction)
print(sample_prediction)
print(sample_score)

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
# Done

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficient = model.coef_
intercept = model.intercept_
print(coefficient)
print(intercept)

[830451.79732027 969401.72391381]
1.0
[ 866.14641337 1032.69506649]
-409391.47958340764


The score for that model is 1.0. This means the model exactly fits the data given. This makes sense, as I am using data from the same place both to predict and to score.

A: Temperature in $^{\circ} C$ \
B: Moles KCl \
h(A,B): Size nm<sup>3</sup> \
Function:  $h(A,B)= 866.14641A + 1032.69507B - 409391.47958$ 


## Part 4. Use Cross Validation

In [56]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = sklearn.model_selection.cross_val_score(model, X_test, Y_test)
print(scores)
print(s.mean(scores))
# Report on their finding and their significance


[0.86600166 0.81565018 0.84209182 0.85686866 0.85851145]
0.8478247528885137


The model seems to have a R<sup>2</sup> value of around 0.85
This means the regression is about an 85% fit to the data.

## Part 5. Using Polynomial Regression

In [57]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly_test = sklearn.preprocessing.PolynomialFeatures(degree = 2).fit(X_test, Y_test)
ptest_data = poly.fit_transform(X_test, Y_test)

poly = sklearn.preprocessing.PolynomialFeatures(degree = 2).fit(X_train, Y_train)
poly_data = poly.fit_transform(X_train, Y_train)
#print(poly_data)
print(poly.powers_)
poly_model = sklearn.linear_model.LinearRegression().fit(poly_data, Y_train)


base_score = poly_model.score(ptest_data, Y_test)
p_scores = sklearn.model_selection.cross_val_score(poly_model, ptest_data, Y_test)
print(base_score)
print(p_scores)
print(s.mean(p_scores))



# Report on the metrics and output the resultant equation as you did in Part 3.
p_coefficients = poly_model.coef_
p_intercept = model.intercept_
#print(p_powers)
print(p_coefficients)
print(p_intercept)


[[0 0]
 [1 0]
 [0 1]
 [2 0]
 [1 1]
 [0 2]]
1.0
[1. 1. 1. 1. 1.]
1.0
[ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
-409391.47958340764


The R^2 score for that model is 1.0, both in simple scoring as well as cross-validated scoring. This means we have identified a function that perfectly matches the data.\
\
The resulting equation is ---------------------------\
A: Temperature in $^{\circ} C$ \
B: Moles KCl \
h(A,B): Size nm<sup>3</sup> \
Function:  $h(A,B)= 12A - 1.27195*10^{-7}*B+1.26494*10^{-11}*A^2+2AB+2.85714*10^{-2}*B^2-409391.47958$
